### PR TITLE
Implement sort on typed list

### PR DESCRIPTION
**Pull request by sklam**
* Sort implementation is borrowed from reflected list.
* supports `key` and `reverse` argument.

Note: 
`sorted()` signature returns a `list`. We can't modify it to return typed list until we allow the `disable_reflected_list` switch.

### CODE REVIEW

File: numba/typed/typedlist.py [hunk 1]
```diff
@@ -299,6 +304,16 @@ def copy(self):
     def index(self, item, start=None, stop=None):
         return _index(self, item, start, stop)
 
+    def sort(self, key=None, reverse=False):
+        """Sort the list inplace.
+
+        See also ``list.sort()``
+        """
+        # If key is not already a dispatcher object, make it so
```
Comment (stuartarchibald) [inline numba/typed/typedlist.py]: This ought to work? But the key isn't a dispatcher type.
```python
from numba import njit
import numpy as np
from numba.typed import List

l = List()
[l.append(x) for x in range(10)]

key = njit(lambda x: 1./x)
l.sort(reverse = True, key=key)
print(l)
```

File: numba/tests/test_typedlist.py [hunk 2]
```diff
@@ -907,3 +907,50 @@ def test_equals_on_list_with_dict_for_unequal_lists(self):
...
+        ]
+        possible_reverse = [True, False]
+        for key, reverse in product(possible_keys, possible_reverse):
+            my_lists = self.make_both(np.random.randint(0, 100, 23))
+            msg = "case for key={} reverse={}".format(key, reverse)
+            self.assertEqual(
+                list(udt(my_lists['nb'], key=key, reverse=reverse)),
+                udt(my_lists['py'], key=key, reverse=reverse),
+                msg=msg,
+            )
```
Comment (stuartarchibald) [inline numba/tests/test_typedlist.py]: Perhaps also test:

1. Use of `typed.List.sort()` inside a jitted function.
2. As 1. but with module scope jitted key function
3. Use of `typed.List.sort()` from the interpreter with an already jitted function as key.

File: numba/typed/typedlist.py (hunk 1 above)
Comment (sklam) [reply numba/typed/typedlist.py]: I mixed up the `numba.types.Dispatcher` and `numba.dispatcher.Dispatcher`.

Comment (sklam) [reply numba/typed/typedlist.py]: fixed now

File: numba/tests/test_typedlist.py (hunk 2 above)
Comment (sklam) [reply numba/tests/test_typedlist.py]: Added tests:
1. `test_sort_in_jit_w_lambda_key`
2. `test_sort_in_jit_w_global_key`
3. `test_sort_dispatcher_key`

**Comment by sklam** [comment]
Testing the impact of zip refleak fix

**Comment by sklam** [comment]
The failures were caused by requirement on scipy.

**Comment by sklam** [comment]
That should be it.

**Pull request merged**